### ISSUE TITLE
[FIX-009] Make non-finite monitor handling explicit before BEST/scheduler selection

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#208]
**Category:** Training and model selection · **Priority:** P1 · **Size:** M

**Evidence status:** Reproduced defect.

## Outcome and scope

Select the first finite value in the existing validation-error, validation-loss, training-error, training-loss order. Warn about rejected non-finite candidates and treat an entirely missing/non-finite signal as unavailable. This chooses the baseline’s fallback policy; do not add named monitor configuration or convert NaN/Inf to a fabricated zero.

## Current behavior / evidence

**Current behavior:** The shared resolver accepts any non-None value, including NaN. Epoch 0 always becomes BEST because there is no previous checkpoint. If its selected error is NaN, all subsequent `finite < NaN` comparisons are false, so BEST freezes on that epoch. The same NaN is supplied to plateau scheduling even when a finite fallback loss is available.

**Concrete reproduction:** Two-epoch tiny CPU run with ordinary finite training and `eval_step_fn` returning `error=nan, loss=1.0` at epoch 0, then `error=0.2, loss=1.0` at epoch 1. Observed `BEST.idp.epoch_idx == 0`, `LAST.idp.epoch_idx == 1`. The first epoch's finite loss is ignored because its error is non-None.

## Repository map

- [`src/nnx/_metrics.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/_metrics.py#L30) — Shared fallback currently tests only `is not None`.
- [`src/nnx/nn/params/nn_run.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py#L166) — Cross-run and within-run comparable score; missing signal maps to positive infinity.
- [`src/nnx/nn/nn_model.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/nn_model.py#L1440) — Initial BEST selection and subsequent comparison.
- [`src/nnx/nn/nn_model.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/nn_model.py#L1837) — Plateau scheduling skips `None` but currently accepts NaN.
- [`src/nnx/trainer/trainer.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/trainer.py#L147) — Multi-optimizer scheduler consumer must share the same policy.
- [`tests/test_metrics.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_metrics.py#L37) — Existing finite priority and all-None unit oracles.
- [`tests/test_eval_step_fn.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_eval_step_fn.py#L36) — Custom per-epoch validation fixture.
- [`tests/test_train_integration.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_train_integration.py#L1194) — Custom-paradigm loss fallback compatibility.

## Implementation approach

1. Walk each EDP’s `error` then `loss` individually, checking `math.isfinite` for each present value. Skip NaN and both infinities; retain `None` as an ordinary absent value. Crucially, an invalid error must still allow the same EDP’s finite loss before moving to training data.
2. Return `None` when no finite signal exists. Keep plateau schedulers from stepping on unavailable values and make warnings identify the rejected field/split; add epoch context at callers where available without storing it in serialized EDPs. Avoid emitting an uncontrolled warning per progress-bar refresh.
3. Route `_best_err` through that resolver so historical NaN BEST checkpoints become replaceable by a finite candidate. Preserve legacy all-None no-signal behavior, but never persist NaN as a comparison baseline. An initial compatibility checkpoint without a signal must compare as unavailable and be replaceable later.
4. Exercise both NNModel and Trainer checkpoint/scheduler consumers. Leave finite ordering, strict less-than improvement, training-loss finite guards, default IDs and evaluation serialization unchanged. Do not reinterpret accuracy/F1 as new maximize-mode monitors.

## Acceptance criteria

- [ ] The reproduction never leaves a NaN-ranked epoch permanently BEST; under the fallback policy epoch 1 replaces epoch 0.
- [ ] NaN, positive infinity and negative infinity have explicit tested handling, including finite loss behind non-finite error.
- [ ] `None` error with finite loss still works for custom paradigms; all-None metrics retain the current supported no-signal behavior.
- [ ] Both NNModel and Trainer scheduler paths avoid feeding non-finite values into metric-driven schedulers and explain skipped/invalid metrics.
- [ ] Existing finite metric ordering and cross-run comparison behavior remain unchanged; legacy NaN-scored BEST artifacts can recover when a finite candidate is saved.

## Validation and completion

Minimal deterministic resolver oracle (an implementation-level test):

```python
from types import SimpleNamespace
from nnx._metrics import _resolve_metric

train = SimpleNamespace(error=.2, loss=.5)
for invalid in (float("nan"), float("inf"), float("-inf")):
    val = SimpleNamespace(error=invalid, loss=1.)
    assert _resolve_metric(val, train) == 1.
    val.loss = invalid
    assert _resolve_metric(val, train) == .2
assert _resolve_metric(SimpleNamespace(error=None, loss=None), None) is None
```

Add `test_resolve_metric_skips_nonfinite_in_priority_order`, `test_nonfinite_first_epoch_best_recovers`, and `test_plateau_never_receives_nonfinite_metric`. In the two-epoch public training fixture return validation `(error=NaN, loss=1)` then `(.2,1)` and assert BEST epoch 1/LAST epoch 1. Load an intentionally legacy NaN-scored BEST and save a finite candidate; verify recovery. Spy on actual scheduler `step` arguments for both trainers, including all-unavailable input (no call), and assert warnings distinguish invalid from absent metrics.

From the repository root, use `uv sync --frozen --all-extras` for the CI-equivalent environment. Existing focused coverage (files already exist):

```bash
uv run --frozen pytest tests/test_metrics.py tests/test_eval_step_fn.py tests/test_train_integration.py tests/test_trainer.py
```

Add the proposed regression cases above to those existing test files, then rerun the same command. Finish with:

```bash
uv run --frozen ruff check src/ tests/ examples/ scripts/
uv run --frozen ruff format --check src/ tests/ examples/ scripts/
uv run --frozen pyright --warnings
```

No new dependency is needed for the implementation described. Preserve Python 3.10 compatibility and default-field omission in existing serialization. These commands are completion requirements for implementation; this specification review does not claim they were run or passed.

## Cascading changes and affected consumers

- **Control signal versus recorded observations.** The finite-only resolver in [_resolve_metric](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/_metrics.py) feeds progress reporting, plateau scheduling, per-run BEST and [global best comparison](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py). Reuse the same chosen value and field/split provenance at those consumers so an invalid validation error followed by finite validation loss does not become a training-loss decision elsewhere. Retain raw NaN/inf observations in live EDPs and checkpoint payloads as diagnostics; do not replace them with zero or a synthetic finite loss. Preserve existing CSV readback, where NaN fields normalize to None, rather than introducing a new encoding. Bound rejection warnings at epoch/caller boundaries rather than repeating them for every progress refresh.

- **Callback and reload boundaries.** [EarlyStopping](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/callbacks.py) resolves its explicitly named field separately, while logging callbacks observe raw epoch data. Document that this repair does not silently change that independent named-monitor contract or create maximize semantics. Test NNModel with its custom eval hook; for Trainer, script the supplied model.evaluate result and training EDPs because Trainer has no eval_step_fn argument. Reload history and a legacy NaN-ranked BEST, then add a finite candidate; comparison must recover with unchanged checkpoint payloads and existing CSV absence semantics. Preserve all-None compatibility checkpoints as unavailable comparison baselines, and keep ordinary non-metric scheduler timing unchanged.

- **Executable invalid-metric case.** [Example 26](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/26_custom_eval_step.py) already demonstrates caller-defined evaluation and is one of the executed smoke examples. Add a separately bounded fixture that returns an invalid error with a finite loss, followed by invalid validation with finite training fallback, an epoch with both train and validation error/loss unavailable, and then a finite improvement. Assert the selected BEST, scheduler call arguments, raw live/checkpoint observations and the existing NaN-to-None CSV readback. Use a tiny loader, scripted eval hook and scripted returned training EDPs; the update loss may stay finite while its reported EDP is unavailable. Update the guide to explain the exact fallback order and show how custom evaluators expose loss-only control signals without mislabelling regression error as classification accuracy.

### Samples and executable coverage

- [`examples/26_custom_eval_step.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/26_custom_eval_step.py): Add a tiny NNModel helper scripting evaluation and returned training EDPs: invalid validation error/finite validation loss; invalid validation with finite train fallback; both train/validation error and loss unavailable; then finite improvement. Assert BEST, scheduler inputs, live/checkpoint diagnostics and existing CSV readback. **Verify:** Proposed execution node: uv run --frozen pytest tests/test_examples_smoke.py -k nonfinite_metric_workflow. CPU, offline, temporary working directory; call the bounded helper rather than the full training main. Retain import-without-main coverage and add the callable to explicit execution coverage; import success alone is insufficient.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | Document finite validation-error → validation-loss → training-error → training-loss fallback, live/checkpoint diagnostic retention, unchanged CSV NaN-to-None semantics, bounded warnings, all-unavailable behavior and distinct explicit EarlyStopping monitor contract. |
| [`docs/quickstart.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/quickstart.md) | `generated/site/Quickstart.md` | `generated/wiki/Quickstart.md` | Explain skipped plateau updates and recoverable unavailable BEST baselines for custom evaluation; preserve finite lower-is-better defaults. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] NNModel eval_step_fn and a Trainer fixture using scripted model.evaluate/training EDPs retain original live/checkpoint observations while displayed/control scores, BEST comparison and plateau inputs consistently use the documented finite fallback.
- [ ] Reloading a legacy NaN-ranked BEST and CSV history retains checkpoint observations and existing CSV NaN-to-None absence semantics; a later finite candidate recovers the selected artifact without a new history encoding or rewritten checkpoint payload.
- [ ] Only an epoch with no finite error/loss in either training or validation skips metric-driven scheduling; invalid validation with finite training still steps on its train fallback, while ordinary schedulers and all-None compatibility artifacts retain their existing behavior.
- [ ] The bounded Example 26 invalid-metric helper executes its scripted epochs and asserts selected BEST and exact scheduler arguments; its explicit callback monitor behavior is documented separately.
- [ ] The issue-specific changes in `docs/concepts.md`, `docs/quickstart.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** Propagation belongs to this fix and existing APIs. Related tickets coordinate shared contracts/files; they are not prerequisites for the bounded repaired path.

Additional nonblocking consumers/coordination: [FIX-008](https://github.com/thekaveh/NNx/issues/207), [FEAT-003](https://github.com/thekaveh/NNx/issues/213), [FEAT-020](https://github.com/thekaveh/NNx/issues/242).

## Delivery order and dependencies

- **Execution order:** 09 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FIX-008](https://github.com/thekaveh/NNx/issues/207), [FEAT-003](https://github.com/thekaveh/NNx/issues/213), [FIX-020](https://github.com/thekaveh/NNx/issues/230). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. No feature dependency; FEAT-003 later makes monitor selection explicit.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** `nnx._metrics._resolve_metric` accepted any non-None value, so a NaN validation error became a permanent BEST (`finite < nan` is always False) and was fed to `ReduceLROnPlateau` even when that epoch carried a finite loss.

**What changed**
- Finite-only fallback (val error → val loss → train error → train loss) in `_resolve_metric`; a non-finite error no longer blocks the same data point's finite loss.
- `_resolve_scheduler_metric` warns once per epoch naming rejected fields and the value used; "every candidate non-finite" and "no metric available" are distinct messages. Both `NNModel` and `Trainer` (resolved once per epoch across all plateau schedulers) skip the step when nothing finite exists.
- `_best_err` inherits the walk, so legacy NaN-ranked BEST artifacts compare as unavailable and are replaced by the next finite candidate; checkpoint payloads and CSV NaN→None readback unchanged.
- Example 26 `nonfinite_metric_workflow()` bounded helper + smoke-harness execution node; docs (`concepts.md` §4.1/§6.2, `quickstart.md` §2.10, examples catalog, API reference, CHANGELOG) updated.

**PRs:** #267 (develop, merged), #268 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1587 passed, 1 skipped, 2 deselected (`test_docs_projection::test_diagram_extraction_balances_nested_svg`, `::test_png_fallback_records_its_svg_source` — pre-existing on develop; libcairo not loadable from this macOS env, unrelated to this change; CI runs them and passed)
- `ruff check` → All checks passed; `ruff format --check` → 252 files already formatted; `pyright --warnings` → 0 errors, 0 warnings
- `build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / docs+wiki projection tests → pass
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #267 and #268.

**Not verifiable locally:** nothing CUDA-specific is touched; the live Pages/wiki publication is left to the `docs.yml` main-push job (deployment skew, if any, is a follow-through item).
